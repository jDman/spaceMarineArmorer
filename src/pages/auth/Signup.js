import React, { Component } from 'react';

import Auth from './Auth';
import Input from '../../components/form/input/Input';
import Button from '../../components/button/Button';

import requiredCheck from '../../utils/requiredCheck';
import minLengthCheck from '../../utils/minLengthCheck';
import classes from './Signup.module.css';

class Signup extends Component {
  state = {
    loading: false,
    signupFormIsValid: false,
    signupForm: {
      userName: {
        elementType: 'input',
        elementConfig: {
          type: 'text',
          name: 'userName',
          placeholder: 'Your username here...',
          required: true,
        },
        value: '',
        valid: false,
        validators: [requiredCheck(null)],
        touched: false,
      },
      email: {
        elementType: 'input',
        elementConfig: {
          type: 'email',
          name: 'email',
          placeholder: 'you@email.com',
          required: true,
        },
        value: '',
        valid: false,
        validators: [requiredCheck(null)],
        touched: false,
      },
      password: {
        elementType: 'input',
        elementConfig: {
          type: 'password',
          name: 'password',
          placeholder: 'Enter you password...',
          required: true,
        },
        value: '',
        valid: false,
        validators: [requiredCheck(null)],
        touched: false,
      },
    },
  };

  inputChangedHandler(value, identifier) {
    this.setState((prevState) => {
      let isValid = true;

      for (const validator of prevState.signupForm[identifier].validators) {
        isValid = isValid && validator(value);
      }

      const updatedForm = {
        ...prevState.signupForm,
        [identifier]: {
          ...prevState.signupForm[identifier],
          touched: true,
          valid: isValid,
          value: value,
        },
      };

      let formIsValid = true;

      for (const inputName in updatedForm) {
        formIsValid = formIsValid && updatedForm[inputName].valid;
      }

      return {
        signupForm: updatedForm,
        signupFormIsValid: formIsValid,
      };
    });
  }

  checkValidity(value, rules) {
    let isValid = true;

    const trimedValue = value.trim();

    if (rules.required) {
      isValid = trimedValue !== '' && trimedValue !== '0' && isValid;
    }

    if (rules.maxLength) {
      isValid = value.length <= rules.maxLength && isValid;
    }

    if (rules.minLength) {
      isValid = value.length >= rules.minLength && isValid;
    }

    return isValid;
  }

  render() {
    const signupFormElements = [];
    const signupForm = this.state.signupForm;

    for (let key in signupForm) {
      signupFormElements.push({
        id: key,
        config: signupForm[key],
      });
    }

    return (
      <Auth>
        <form onSubmit={(event) => this.props.signup(event, this.state)}>
          {signupFormElements.map((formElement) => {
            return (
              <Input
                key={formElement.id}
                elementType={formElement.config.elementType}
                elementConfig={formElement.config.elementConfig}
                value={formElement.config.value}
                invalid={!formElement.config.valid}
                touched={formElement.config.touched}
                hasValidation={formElement.config.validators.length}
                changed={(event) =>
                  this.inputChangedHandler(event.target.value, formElement.id)
                }
              />
            );
          })}
          <Button
            type="submit"
            classes={
              !this.state.signupFormIsValid
                ? [classes.SignupFormButton, classes.SignupFormButtonDisabled]
                : classes.SignupFormButton
            }
            buttonText="Signup"
            disabled={!this.state.signupFormIsValid}
          />
        </form>
      </Auth>
    );
  }
}

export default Signup;
