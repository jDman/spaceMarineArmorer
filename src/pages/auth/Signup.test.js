import React from 'react';
import renderer from 'react-test-renderer';

import Signup from './Signup';

describe('<Signup />', () => {
  it('renders Signup page', () => {
    const component = renderer.create(<Signup />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates email form state on input change', () => {
    const component = renderer.create(<Signup />);

    const componentInstance = component.root;

    componentInstance
      .findByProps({ name: 'email' })
      .props.onChange({ target: { value: 'test@test.com' } });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates password form state on input change', () => {
    const component = renderer.create(<Signup />);

    const componentInstance = component.root;

    componentInstance
      .findByProps({ name: 'password' })
      .props.onChange({ target: { value: '12345678' } });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates userName form state on input change', () => {
    const component = renderer.create(<Signup />);

    const componentInstance = component.root;

    componentInstance
      .findByProps({ name: 'userName' })
      .props.onChange({ target: { value: 'fred1' } });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
