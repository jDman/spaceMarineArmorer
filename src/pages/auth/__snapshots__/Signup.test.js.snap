// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders Signup page 1`] = `
<section
  className="auth-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="userName"
        onChange={[Function]}
        placeholder="Your username here..."
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="email"
        onChange={[Function]}
        placeholder="you@email.com"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="password"
        onChange={[Function]}
        placeholder="Enter you password..."
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Signup
    </button>
  </form>
</section>
`;

exports[`<Signup /> updates email form state on input change 1`] = `
<section
  className="auth-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="userName"
        onChange={[Function]}
        placeholder="Your username here..."
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="email"
        onChange={[Function]}
        placeholder="you@email.com"
        required={true}
        type="email"
        value="test@test.com"
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="password"
        onChange={[Function]}
        placeholder="Enter you password..."
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Signup
    </button>
  </form>
</section>
`;

exports[`<Signup /> updates password form state on input change 1`] = `
<section
  className="auth-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="userName"
        onChange={[Function]}
        placeholder="Your username here..."
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="email"
        onChange={[Function]}
        placeholder="you@email.com"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="password"
        onChange={[Function]}
        placeholder="Enter you password..."
        required={true}
        type="password"
        value="12345678"
      />
    </div>
    <button
      type="submit"
    >
      Signup
    </button>
  </form>
</section>
`;

exports[`<Signup /> updates userName form state on input change 1`] = `
<section
  className="auth-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="userName"
        onChange={[Function]}
        placeholder="Your username here..."
        required={true}
        type="text"
        value="fred1"
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="email"
        onChange={[Function]}
        placeholder="you@email.com"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        name="password"
        onChange={[Function]}
        placeholder="Enter you password..."
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Signup
    </button>
  </form>
</section>
`;
